Add design goals for L1 contracts (#4)
ethereum-optimism/optimistic-specs: layer-1-contracts/README.md

# Layer 1 Contracts

## Design Goals
We want our ORU contracts to be built in compliance with the following goals:
- A **pluggable state machine** which:
    - allows separation of concerns and parallelizable work

>     - is fully stateless, i.e. all state transitions can be expressed in the form `bytes32 _inputHash, bytes _witness => bytes32 _outputhash` and are only dependent on the immediately preceding pre-state

Author: cc @protolambda , this is an explicit callout that steps are only dependent on the immediately preceding one

- A cryptoeconomic **State Oracle** which:
    - Provides a cryptoeconomic mechanism to propose, assert, challenge, and finalize optimistic states of the L2 machine
        - Does NOT require that the entire execution trace is provable, just the execution output

>             - i.e. an honest party may not necessarily ever post a hash of the full trace

Author: This callout is one of the assumptions allowing #3 [here](https://statechannels.notion.site/Various-bisection-games-8d5ed428f7114b52a1c6bbeb08fe96d5)

Reviewer: FYI `#3` links to pull request #3 in this repo, which is misleading. I think you mean "Option 3" in the linked notion doc, right?

>         - DOES require that there is a winning strategy for all, and only for all, possible state proposals

Reviewer: Should this say something like this?
```suggestion
        - DOES require that there is a winning strategy for correct state proposals
```

I'm not sure what guarantees you can make in case BOTH parties are making incorrect state proposals. But I don't know if you care about that case?

Author: Used your suggested wording, and added a bullet below it with the (probably obvious but hey) thing I was trying to convey with "only for all"
